app/server/lib/FlexServer.ts

```diff
@@ -405,24 +405,31 @@ export class FlexServer implements GristServer {
 
   public addLogging() {
     if (this._check('logging')) { return; }
-    if (process.env.GRIST_LOG_SKIP_HTTP) { return; }
+    if (isAffirmative(process.env.GRIST_LOG_SKIP_HTTP)) { return; }
     // Add a timestamp token that matches exactly the formatting of non-morgan logs.
     morganLogger.token('logTime', (req: Request) => log.timestamp());
     // Add an optional gristInfo token that can replace the url, if the url is sensitive.
     morganLogger.token('gristInfo', (req: RequestWithGristInfo) =>
                        req.gristInfo || req.originalUrl || req.url);
     morganLogger.token('host', (req: express.Request) => req.get('host'));
-    const msg = ':logTime :host :method :gristInfo :status :response-time ms - :res[content-length]';
+    morganLogger.token('body', (req: express.Request) => JSON.stringify(req.body));
+
+    // For debugging, be careful not to enable logging in production (may log sensitive data)
+    const shouldLogBody = isAffirmative(process.env.GRIST_LOG_HTTP_BODY);
+
+    const msg = `:logTime :host :method :gristInfo ${shouldLogBody ? ':body' : ''} ` +
+      ":status :response-time ms - :res[content-length]";
     // In hosted Grist, render json so logs retain more organization.
     function outputJson(tokens: any, req: any, res: any) {
       return JSON.stringify({
         timestamp: tokens.logTime(req, res),
+        host: tokens.host(req, res),
         method: tokens.method(req, res),
         path: tokens.gristInfo(req, res),
+        body: shouldLogBody ? tokens.body(req, res) : {},
```

Comment:
Hmm, I don't know if this is intentional, but it looks to me like this will log `{ … path: "/whatever", body: {}, status: 200, … }` when body logging is disabled, and `{ … path: "/whatever", body: "body-content", status: 200, … }` when enabled. Having `body` switch from being a string to an (empty) object doesn't seem correct.

My earlier suggestion was to use object spreading to make `body` go away entirely when not enabled:

```suggestion
        ...(shouldLogBody ? { body: tokens.body(req, res) } : {}),
```

Reply:
You're right, thanks!

Reply:
That being said, while it should work when logging in JSON, it still prints `{}` when logging in plain text:

```
2024-03-28 18:20:45.193 localhost:8484 GET /o/docs/api/orgs/0/workspaces?includeSupport=1 {} 304 39.934 ms - -
```

Reply:
Err, this is quite weird, how would a change to the JSON logging impact the plain text logger?

Reply:
It doesn't, it seems like the JSON and the plain text logger are different things.

But whereas what you suggested removes the body when we enable logging in JSON, it still prints `{}` when the plain text logging is enabled.

Unless we find a way to skip the body here:
https://github.com/gristlabs/grist-core/pull/913/files#diff-cc10ec0f715f5f40f94fcf195753f0052aede521998746b91689ce96b9a30055R420-R421

If I am correct, Nginx in this case prints a `-` when no values exist, maybe we could do the same?

Reply:
Here's what I'm observing when running your branch (with `GRIST_LOG_SKIP_HTTP=false` and `GRIST_HOSTED_VERSION` unset in all cases), when doing a `POST` to `/` with body `{"x":1}` :
- with `GRIST_LOG_HTTP_BODY=false`:
```
localhost:8484 POST /  404 3.977 ms - 2347
```
No `{}` here, but note that there's a double space here between `/` and `404`, which you could get rid of by tweaking the template to include the space before `:body` inside the conditional.
- with `GRIST_LOG_HTTP_BODY=true GRIST_HOSTED_VERSION=`:
```
localhost:8484 POST / {"x":"1"} 404 12.982 ms - 2347
```
The body is correctly printed here. However if I do a request without a body, e.g. `GET /` this is printed:
```
localhost:8484 GET / {} 302 8.375 ms - 51
```
In fact, if I do a POST with a non-JSON request body I also get `{}` printed. And `"body": "{}"` if I enable JSON logging with `GRIST_HOSTED_VERSION=true` (btw that variable might also benefit from consistent application of `isAffirmative` — it is referenced three times through the code, but only once with `isAffirmative`).

Maybe `JSON.stringify(req.body)` is not be the best way to get at the body for logging? Or maybe it's fine since there's not much point in logging non-JSON bodies?

Reply:
I think I have come up with a solution for your remark here, what do you think?
https://github.com/gristlabs/grist-core/pull/913/commits/90358596429f90e5102620f52cab82091a48022c